prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"path"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// A fetcher for a GCS client
+type GCSArtifactFetcher struct {
+	client *storage.Client
+}
+
+// A location in GCS where prow job-specific artifacts are stored
+type GCSJobSource struct {
+	source     string
+	linkPrefix string
+	bucket     string
+	jobPath    string
+}
+
+// NewGCSArtifactFetcher creates a new ArtifactFetcher with a real GCS Client
+func NewGCSArtifactFetcher() *GCSArtifactFetcher {
+	c, err := storage.NewClient(context.Background(), option.WithoutAuthentication())
+	if err != nil {
+		log.Fatal(err)
+	}
+	return &GCSArtifactFetcher{
+		client: c,
+	}
+}
+
+// NewGCSJobSource creates a new GCSJobSource from a given bucket and jobPath
+func NewGCSJobSource(src string) *GCSJobSource {
+	linkPrefix := "gs://"
+	noPrefixSrc := strings.TrimPrefix(src, linkPrefix)
+	bucket := strings.Split(noPrefixSrc, "/")[0]
+	jobPath := strings.TrimPrefix(noPrefixSrc, bucket+"/")
+	return &GCSJobSource{
+		source:     src,
+		linkPrefix: linkPrefix,
+		bucket:     bucket,
+		jobPath:    jobPath,
+	}
+}
+
+// Artifacts gets all artifacts from a GCS job source
```

Comment:
As mentioned on the design doc, why do we want this to be synchronous? It seems like the biggest issue with larger job buckets is going to be a bottleneck as we collect _every_ artifact and render them all. Can we stream instead?

Reply:
it's not going to be synchronous and already the rendering is not, the view page loads each view async.  IIRC this code just creates handles and is undergoing refactor currently.

Reply:
more concretely a generated view page without any rendered contents is the page load, then javascript fetches the build log view, junit, etc. all of the GCS content streaming should be handled per view

Reply:
To elaborate further, we check which artifacts are available in the job bucket first so we can pre-render the view containers. Each view is then loaded async. In each of these async calls we get concrete handles to artifacts relevant only to that particular viewer. It is the viewer implementation's responsibility to perform any read operations on the artifact. The only synchronous call here is listing the contents of the bucket, which is necessary for the page to know which viewers to pre-render and load.